File: 1763. Longest Nice Substring/optimised.ts
function longestNiceSubstring(s: string): string {
	if (s.length < 2) return "";
	
	const chars = new Set(s);

	for (let i = 0; i < s.length; i++) {
		const c = s[i]
		const op = c === c.toLowerCase() ? c.toUpperCase() : c.toLowerCase(); 

		if (!chars.has(op)) {
			const left = longestNiceSubstring(s.substring(0, i));
			const right = longestNiceSubstring(s.substring(i + 1));

			return left.length >= right.length ? left : right;
		}
	}

	return s
}

console.log(longestNiceSubstring("YazaAay"));

File: 1763. Longest Nice Substring/optimised.py
#!/usr/bin/env python

def longestNiceSubstring(s: str) -> str:
    if len(s) < 2:
        return ""

    chars = set(s)

    for i in range(len(s)):
        c = s[i]
        op = c.swapcase()

        if op not in chars:
            left = longestNiceSubstring(s[:i])
            right = longestNiceSubstring(s[i + 1:])

            return left if len(left) >= len(right) else right

    return s

print(longestNiceSubstring("YazaAay"))
